### PR TITLE
fix: use a single client for upgrades

### PR DESCRIPTION
**Pull request by andrewrynhard**
Without this change, the upgrade process creates a new client for every
reconciliation. This was a problem for two reasons. First, there was no
way to close the client connection once reconcilitation was done, so
connections were left open, and exhausted trustd's FD limit. This
has since been fixed in upstream Talos. The second issue is not really
an "issue", but more of an optimization. There really is no need to
create a client every time. We could create one client, that
automatically renew's it's certificate, and adjusts the targeted node by
setting it in the request's context. This is what this change does.

### CODE REVIEW

**Comment by smira** [comment]
I guess with need client-side load-balancing across endpoints (https://github.com/talos-systems/talos/issues/1641) to make this smooth, as when 1st master node goes down, upgrader is going to become blind.

File: go.mod [hunk 1]
```diff
@@ -2,6 +2,8 @@ module github.com/talos-systems/talos-controller-manager
 
 go 1.13
 
+replace github.com/talos-systems/talos => /vagrant/code/github.com/talos-systems/talos
```
Comment (smira) [inline go.mod]: looks like this shouldn't be here ;)

**Comment by andrewrynhard** [comment]
> I guess with need client-side load-balancing across endpoints ([talos-systems/talos#1641](https://github.com/talos-systems/talos/issues/1641)) to make this smooth, as when 1st master node goes down, upgrader is going to become blind.

Yeah, and I discovered that the mutual TLS setup for this might be interesting to implement with client-side load balancing too. In the TLS config we need to set `ServerName`, but that will change depending on with endpoint we are talking to. Will be a fun problem to solve.

File: go.mod (hunk 1 above)
Comment (andrewrynhard) [reply go.mod]: good catch ... updated to use my fork until the PR in Talos is merged.

File: pkg/upgrader/upgrader_v1alpha1.go [hunk 2]
```diff
@@ -36,99 +35,107 @@ import (
...
+
+	ca, err := certificateProvider.GetCA()
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("failed to get root CA: %w", err)
 	}
 
-	ca, crt, err := generator.Identity(csr)
+	creds, err := tls.New(
+		tls.WithClientAuthType(tls.Mutual),
```
Comment (andrewrynhard) [inline pkg/upgrader/upgrader_v1alpha1.go]: I'm not sure we need to actually set this for the client TLS config.

**Pull request merged**